hvplot/plotting/scatter_matrix.py

```diff
@@ -38,19 +47,99 @@ def scatter_matrix(data, c=None, chart='scatter', diagonal='hist', alpha=0.5, **
+                "Spread will not be applied to plots."
+                )
+        sp_kwds = {}
+        if 'px' in kwds:
+            sp_kwds['px'] = kwds.pop('px')
+        if 'shape' in kwds:
+            sp_kwds['shape'] = kwds.pop('shape')
+        if 'how' in kwds:
+            sp_kwds['how'] = kwds.pop('how')
+        if 'mask' in kwds:
+            sp_kwds['mask'] = kwds.pop('mask')
+            
+        
+
     colors = _hv.plotting.util.process_cmap('Category10', categorical=True)
     chart_opts = dict(alpha=alpha, cmap=colors, tools=['box_select', 'lasso_select'],
                       nonselection_alpha=0.1, **kwds)
 
+    #get initial scatter matrix.  No color.
     grid = _hv.operation.gridmatrix(data, diagonal_type=diagonal, chart_type=chart)
+
     if c:
+        #change colors for scatter matrix
         chart_opts['color_index'] = c
+        # not entirely sure what the next line is doing, but its helping
+        # with the coloring.
         grid = grid.map(lambda x: x.clone(vdims=x.vdims+[c]), 'Scatter')
+        # create a new scatter matrix with groups for each catetory, so now the histogram will
+        # show separate colors for each group.
         groups = _hv.operation.gridmatrix(data.groupby(c).overlay(),
                                           chart_type=chart,
                                           diagonal_type=diagonal)
+        # take the correct layer from each Overlay object within the scatter matrix.
         grid = (grid * groups).map(lambda x: x.get(0) if isinstance(x.get(0), chart) else x.get(1),
                                    _hv.Overlay)
 
+    # set the histogram colors 
     diagonal_opts = {'fill_color': _hv.Cycle(values=colors)}
-    return grid.options({chart.__name__: chart_opts, diagonal.__name__: diagonal_opts})
+    # actually changing to the same color scheme for both scatter and histogram plots.
+    grid = grid.options({chart.__name__: chart_opts, diagonal.__name__: diagonal_opts})
+    
+    # Perform datashade options after all the coloring is finished.
+    if datashade or rasterize:
+        if dynspread:
+            grid = grid.map(lambda x: x.apply(hd.datashade if datashade 
+                                            else hd.rasterize, **ds_kwds)
+                                        .apply(hd.dynspread, **dn_kwds)
+                                        if isinstance(x, chart) else x)
+        elif spread:
+            grid = grid.map(lambda x: x.apply(hd.datashade if datashade 
+                                            else hd.rasterize, **ds_kwds)
+                                        .apply(hd.spread, **sp_kwds)
+                                        if isinstance(x, chart) else x)
+        else:
+            grid = grid.map(lambda x: x.apply(hd.datashade if datashade 
+                                            else hd.rasterize, **ds_kwds) 
+                            if isinstance(x, chart) else x)
```

Comment:
Seems like this can all be done just once:

```suggestion
    if datashade or rasterize:
        aggregatefn = hd.datashade if datashade else hd.rasterize
        spreadfn    = hd.dynspread if dynspread else (hd.spread if spread else lambda z, **_: z)
    
        grid = grid.map(lambda x: x.apply(aggregatefn, **ds_kwds).apply(spreadfn, **sp_kwds) if isinstance(x, chart) else x)
```
It's slightly tricky to have to define a no-op when there is no spreading, but I still think it's better than copying all that grid.map code with subtle differences.